src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java

```diff
@@ -0,0 +1,335 @@
+                        workflowId,
+                        deprovisionStepId
+                    ),
+                    Collections.emptyList(),
+                    this.threadPool,
+                    pn.nodeTimeout()
+                );
+            })
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        // Deprovision in reverse order of provisioning to minimize risk of dependencies
+        Collections.reverse(deprovisionProcessSequence);
+        logger.info("Deprovisioning steps: {}", deprovisionProcessSequence.stream().map(ProcessNode::id).collect(Collectors.joining(", ")));
+
+        // Repeat attempting to delete resources as long as at least one is successful
+        int resourceCount = deprovisionProcessSequence.size();
+        while (resourceCount > 0) {
+            Iterator<ProcessNode> iter = deprovisionProcessSequence.iterator();
+            while (iter.hasNext()) {
+                ProcessNode deprovisionNode = iter.next();
+                ResourceCreated resource = getResourceFromDeprovisionNode(deprovisionNode, resourceMap);
+                String resourceNameAndId = getResourceNameAndId(resource);
+                CompletableFuture<WorkflowData> deprovisionFuture = deprovisionNode.execute();
+                try {
+                    deprovisionFuture.join();
+                    logger.info("Successful {} for {}", deprovisionNode.id(), resourceNameAndId);
+                    // Remove from list so we don't try again
+                    iter.remove();
+                    // Pause briefly before next step
+                    Thread.sleep(100);
+                } catch (Throwable t) {
+                    logger.info(
+                        "Failed {} for {}: {}",
+                        deprovisionNode.id(),
+                        resourceNameAndId,
+                        t.getCause() == null ? t.getMessage() : t.getCause().getMessage()
+                    );
+                }
+            }
+            if (deprovisionProcessSequence.size() < resourceCount) {
+                // If we've deleted something, decrement and try again if not zero
+                resourceCount = deprovisionProcessSequence.size();
+                deprovisionProcessSequence = deprovisionProcessSequence.stream().map(pn -> {
+                    return new ProcessNode(
+                        pn.id(),
+                        workflowStepFactory.createStep(pn.workflowStep().getName()),
+                        pn.previousNodeInputs(),
+                        pn.input(),
+                        pn.predecessors(),
+                        this.threadPool,
+                        pn.nodeTimeout()
+                    );
+                }).collect(Collectors.toList());
+                // Pause briefly before next loop
+                try {
+                    Thread.sleep(1000);
```

Comment:
Also why pause here as well? Just trying to understand better

Reply:
This is primarily for the "delete model" step that somehow doesn't propogate enough to allow deleting the connector, without at least 500ms but something less than 1s delay.  

Unlike deploying models where we can get a task ID and check for completion, the deletions return immediately (with success) but we have to rely on normal OpenSearch node communication/replication to share the state around.

Reply:
To summarize the delays (with local testing)

Reply:
To summarize (with local testing):
1. With no delay: occasional failure after "undeploy" that prevents "delete".  Any short delay resolves. This is probably a "local node" race condition.
2. With no delay: always failure after "delete model" that prevents "delete connector".  Delay over 500ms required.  Delay of 1s alway works. This is probably a "replicate state to all nodes" race condition.